node/core/backing/src/lib.rs

```diff
@@ -0,0 +1,1149 @@
+struct TableContext {
+	signing_context: SigningContext,
+	key: Option<Arc<ValidatorPair>>,
+	groups: HashMap<ParaId, HashSet<ValidatorId>>,
+	validators: Vec<ValidatorId>,
+}
+
+impl TableContextTrait for TableContext {
+	fn is_member_of(&self, authority: ValidatorIndex, group: &ParaId) -> bool {
+		let key = match self.validators.get(authority as usize) {
+			Some(val) => val,
+			None => return false,
+		};
+
+		self.groups.get(group).map_or(false, |g| g.get(&key).is_some())
+	}
+
+	fn requisite_votes(&self, group: &ParaId) -> usize {
+		self.groups.get(group).map_or(usize::max_value(), |g| needed_votes(g.len()))
+	}
+}
+
+impl TableContext {
+	fn local_id(&self) -> Option<ValidatorId> {
+		self.key.as_ref().map(|k| k.public())
+	}
+
+	fn local_index(&self) -> Option<ValidatorIndex> {
+		self.local_id().and_then(|id|
+			self.validators
+				.iter()
+				.enumerate()
+				.find(|(_, k)| k == &&id)
+				.map(|(i, _)| i as ValidatorIndex)
+		)
+	}
+}
+
+const CHANNEL_CAPACITY: usize = 64;
+
+/// A message type that is sent from `CandidateBackingSubsystem` to `CandidateBackingJob`.
+enum ToJob {
+	/// A `CandidateBackingMessage`.
+	CandidateBacking(CandidateBackingMessage),
+	/// Stop working.
+	Stop,
+}
+
+/// A message type that is sent from `CandidateBackingJob` to `CandidateBackingSubsystem`.
+enum FromJob {
+	RuntimeApiMessage(RuntimeApiMessage),
+	AvailabilityStoreMessage(AvailabilityStoreMessage),
+	CandidateValidation(CandidateValidationMessage),
+	StatementDistribution(StatementDistributionMessage),
+	Provisioner(ProvisionerMessage),
+}
+
+impl From<FromJob> for AllMessages {
+	fn from(f: FromJob) -> Self {
+		match f {
+			FromJob::RuntimeApiMessage(msg) => AllMessages::RuntimeApi(msg),
+			FromJob::AvailabilityStoreMessage(msg) => AllMessages::AvailabilityStore(msg),
+			FromJob::CandidateValidation(msg) => AllMessages::CandidateValidation(msg),
+			FromJob::StatementDistribution(msg) => AllMessages::StatementDistribution(msg),
+			FromJob::Provisioner(msg) => AllMessages::Provisioner(msg),
+		}
+	}
+}
+
+// It looks like it's not possible to do an `impl From` given the current state of
+// the code. So this does the necessary conversion.
+fn primitive_statement_to_table(s: &SignedFullStatement) -> TableSignedStatement {
+	let statement = match s.payload() {
+		Statement::Seconded(c) => TableStatement::Candidate(c.clone()),
+		Statement::Valid(h) => TableStatement::Valid(h.clone()),
+		Statement::Invalid(h) => TableStatement::Invalid(h.clone()),
+	};
+
+	TableSignedStatement {
+		statement,
+		signature: s.signature().clone(),
+		sender: s.validator_index(),
+	}
+}
+
+// finds the first key we are capable of signing with out of the given set of validators,
+// if any.
+fn signing_key(validators: &[ValidatorId], keystore: &KeyStorePtr) -> Option<Arc<ValidatorPair>> {
+	let keystore = keystore.read();
+	validators.iter()
+		.find_map(|v| {
+			keystore.key_pair::<ValidatorPair>(&v).ok()
+		})
+		.map(|pair| Arc::new(pair))
```

Comment:
is the `Arc` cloned somewhere? I would imagine that the job can actually maintain full ownership of the key